### ISSUE TITLE
Automatic Witgen: Introduce a notion of bus interaction

### ISSUE THREAD

**Issue by georgwiese** [powdr-labs/powdr#1829]
### Background

Automatic witness generation is currently aware of native lookup and permutation arguments, which can be used to connect otherwise independently constrained subsets of columns.

For example, we might have a lookup like:
`main.instr_xor [ arg1, arg2, out ] in xor.LATCH [ xor.a, xor.b, xor.c ]`

In any given row in `main`, if the `main.instr_xor` flag is active, automatic witgen "knows" that it should "call" the XOR machine to fill any unknown values on the LHS of the lookup.

### Bus interaction

A *bus interaction* (typically thought of as bus send & bus receive) is a generalization of this. In the bus abstraction, tuples can be *sent* to the bus with some multiplicity (typically 0 or 1) and *received* from the bus (a "receive" is a "send" with negative multiplicity). The bus argument ensures that at the very end of the computation, the multi-set of tuples sent equals the multi-set of tuples received.

This can be used be used to implement permutation, as done in our standard library:
https://github.com/powdr-labs/powdr/blob/5739ebcee8773f9d5f6b6e9bcd65d01d6ab320f5/std/protocols/permutation_via_bus.asm#L13-L17
Similarly, a lookup can be implemented.

However, the send / receive paradigm is more general. For example:
- Multiple receives to the same target can be summarized. For example, if $n$ columns are byte-constrained as `[ col ] in [ BYTES ]`, we'd need $n$ sends, but only one receive (and consequently only one multiplicity column that needs to be committed). This is one of the main advantages of [LogUp](https://eprint.iacr.org/2023/1284).
- A send might not be statically connected to a receive; the receiver might depend on some run-time arguments. The most extreme example I've seen of this is the [Ceno zk-VM](https://georgwiese.github.io/crypto-summaries/Papers/Ceno---Non-uniform,-Segment-and-Parallel-Zero-knowledge-Virtual-Machine): Here, each instruction *receives* a $(pc, time\\_step)$ tuple, asserts that the instruction at the provided PC is the one that it is responsible for, runs the instruction, and then sends a new $(pc', time\\_step')$ tuple. Because of the assertion, there can only be one receiver for any given tuple, but who that is depends on the run-time value of the PC (and the program).

### An algorithm to handle arbitrary bus interaction

To simplify slightly, I think we can always assume the following:
- Bus sends are with multiplicity 0 or 1.
- For each send, *exactly one* receive should match *at runtime*.

One possible algorithm to solve this would be the following:
- Keep a list of all bus interactions, i.e. all bus sends and bus receives (these should come from witgen annotations, see #1378)
- Whenever a send is active, there needs to be a matching receive, we might not know which. So, we can:
  - Try all receives (in parallel?)
  - Complete if exactly one "completes" (in the same sense as today a [`process_plookup`](https://github.com/powdr-labs/powdr/blob/5739ebcee8773f9d5f6b6e9bcd65d01d6ab320f5/executor/src/witgen/machines/mod.rs#L52) call can complete)

For this to work we would need a witgen annotation (#1378) containing the following information:
- Whether it's a send or receive. A send would always have a multiplicity of 0 or 1; a receive would always have negative multiplicity. Receives should only be used to match with sends.
- An expression for the multiplicity. This activates a send or tells witgen the multiplicity column in the case of a receive.
- The tuple being sent / received.

**Comment by georgwiese** [comment]
I think the algorithm described above can be simplified if we assume that the first entry of the sent / received tuple is an ID, which is static in receives. All use-cases I'm aware of are like that.

To spec out the algorithm a bit more:
- Instead of phantom lookups / permutations, we would have bus receive / send annotation (specified above, but maybe an extra field to name the ID explicitly instead of relying on the first element in the tuple).
- The machine extractor needs to be rewritten to work on bus sends (treated like lookup / permutation LHSs now) and bus receives (treated like lookup / permutation RHSs now).
- Native lookups / permutations can be translated to equivalent bus annotations for that purpose.
  - The LHS turns into a send; the RHS turns into a receive
  - The tuple being sent is the tuple in the LHS / RHS
  - Their bus ID would be the identity ID
  - The LHS / RHS selector would be the multiplicity
    - *TODO: This is not correct for the RHS of a lookup*
- Instead of [mapping from identity ID to machine](https://github.com/powdr-labs/powdr/blob/ff054afd454d5a93c85546d76fc0e6445524e003/executor/src/witgen/identity_processor.rs#L25), we can map from bus ID to machine.

**Comment by leonardoalt** [comment]
@georgwiese I guess this is done?

**Comment by georgwiese** [comment]
No, this is about bus interactions in the general version. E.g. situations where you send a tuple but don't know the receiver at compile time. For that, witgen is not yet supported.
